### PR TITLE
Add dumping of reproducer for toy sharded llama test

### PR DESCRIPTION
**Pull request by sogartar**
Add pytest CLI option for overriding the temporary test directory. If the option is provided dump MLIR, VMFB, function call arguments and results. Move numerical checks at the end of the test.
Add some basic functionality to run multi-device IREE functions with tracing.

### CODE REVIEW

**Comment by sogartar** [comment]
The CI needs some updates to the `iree-turbine` packages.

File: sharktank/tests/models/llama/sharded_llama_test.py [hunk 1]
```diff
@@ -346,91 +286,105 @@ def _(model, *args, **kwargs) -> torch.Tensor:
...
+            vm_context=vm_context,
+            driver=iree_driver,
+            trace_path_prefix=path_prefix if dump_enabled else None,
+        )
+        decode_iree_result = iree_to_torch(*decode_iree_result)
+        expected_decode_result = call_torch_module_function(
+            module=sharded_model,
+            function_name="decode",
+            kwargs=sharded_decode_args,
+            trace_path_prefix=f"{path_prefix}expected_" if dump_enabled else None,
```
Comment (rsuderman) [inline sharktank/tests/models/llama/sharded_llama_test.py:359]: You should use `os.path.join` for combining

File: sharktank/tests/models/llama/sharded_llama_test.py [hunk 2]
```diff
@@ -346,91 +286,105 @@ def _(model, *args, **kwargs) -> torch.Tensor:
...
+        prefill_iree_result = iree_to_torch(*prefill_iree_result)
+        assert len(prefill_iree_result) == 1
+        expected_prefill_result = call_torch_module_function(
+            module=sharded_model,
+            function_name="prefill",
+            kwargs=sharded_prefill_args,
+            trace_path_prefix=f"{path_prefix}expected_" if dump_enabled else None,
+        )
+        prefill_iree_cache_state_shards = prefill_iree_args[
+            -self.config.tensor_parallelism_size - 1 :
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/tests/models/llama/sharded_llama_test.py:336]: The last `self.config.tensor_parallelism_size` number of arguments is the cache state.

File: sharktank/tests/models/llama/sharded_llama_test.py (hunk 1 above)
Comment (sogartar) [reply sharktank/tests/models/llama/sharded_llama_test.py:359]: This prefix is interpreted as a string, not as a directory. This allows to prefix not only with a directory, but a part of the filename as well.

**Pull request merged**